File: foundation/foundation-transform/aissemble-foundation-transform-core-python/src/data_transform_core/policy/data_transform_policy_manager.py
# #%L
# Data Transform::Python::Core
# %%
# Copyright (C) 2021 Booz Allen
# %%
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#      http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
# #L%
###
from typing import List
from krausening.logging import LogManager
from policy_manager import AbstractPolicyManager
from policy_manager.policy import Policy
from policy_manager.policy.json import PolicyInput, PolicyRuleInput
from data_transform_core.mediator import MediationConfiguration, MediationProperty
from data_transform_core.policy import DataTransformPolicy
from data_transform_core import DataTransformException


class DataTransformPolicyManager(AbstractPolicyManager):
    """
    DataTransformPolicyManager class that overrides the methods in the
    policy manager to add custom configurations for the data transform policies.
    """

    __logger = LogManager.get_instance().get_logger("DataTransformPolicyManager")
    __instance = None

    def __init__(self):
        if DataTransformPolicyManager.__instance is not None:
            raise Exception("Class is a singleton")
        else:
            super().__init__()
            DataTransformPolicyManager.__instance = self

    @staticmethod
    def getInstance():
        if DataTransformPolicyManager.__instance is None:
            DataTransformPolicyManager()
        return DataTransformPolicyManager.__instance

    def createPolicy(self, policyIdentifier: str) -> DataTransformPolicy:
        return DataTransformPolicy(identifier=policyIdentifier)

    def setAdditionalConfigurations(self, policy: Policy, input: PolicyInput) -> None:
        if not isinstance(policy, DataTransformPolicy):
            raise DataTransformException("Policy was not configured for data transform")

        mediationManager = policy.mediationManager
        mediationConfigurations = self.createMediationConfigurations(input.rules)

        for mediationConfiguration in mediationConfigurations:
            mediationManager.validateAndAddMediator(
                mediationConfigurations, mediationConfiguration
            )

    def createMediationConfigurations(
        self, ruleInputs: List[PolicyRuleInput]
    ) -> List[MediationConfiguration]:
        mediationConfigurations = []
        for ruleInput in ruleInputs:
            mediationConfiguration = self.createMediationConfiguration(ruleInput)
            mediationConfigurations.append(mediationConfiguration)

        return mediationConfigurations

    def createMediationConfiguration(
        self, ruleInput: PolicyRuleInput
    ) -> MediationConfiguration:
        mediationConfiguration = None
        try:
            className = ruleInput.className
            inputType = ruleInput.configurations.get("inputType")
            outputType = ruleInput.configurations.get("outputType")
            properties = ruleInput.configurations.get("properties")

            mediationConfiguration = MediationConfiguration(
                className=className, inputType=inputType, outputType=outputType
            )
            if properties:
                mediationConfiguration.properties = [
                    MediationProperty(
                        key=property.get("key"), value=property.get("value")
                    )
                    for property in properties
                ]

            DataTransformPolicyManager.__logger.debug(
                "Created mediation configuration: %s" % mediationConfiguration
            )

        except Exception:
            raise DataTransformException("Invalid configurations found in rule")

        return mediationConfiguration

File: foundation/foundation-transform/foundation-transform-core-java/src/main/java/com/boozallen/data/transform/policy/DataTransformPolicyManager.java
package com.boozallen.data.transform.policy;

/*-
 * #%L
 * aiSSEMBLE::Foundation::Transform::Java
 * %%
 * Copyright (C) 2021 Booz Allen
 * %%
 * This software package is licensed under the Booz Allen Public License. All Rights Reserved.
 * #L%
 */

import java.util.ArrayList;
import java.util.List;

import org.technologybrewery.mash.BaseMediationManager;
import org.technologybrewery.mash.MediationConfiguration;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import com.boozallen.aissemble.core.policy.configuration.policy.DefaultPolicy;
import com.boozallen.aissemble.core.policy.configuration.policy.Policy;
import com.boozallen.aissemble.core.policy.configuration.policy.json.PolicyInput;
import com.boozallen.aissemble.core.policy.configuration.policy.json.rule.PolicyRuleInput;
import com.boozallen.aissemble.core.policy.configuration.policymanager.AbstractPolicyManager;
import com.boozallen.data.transform.DataTransformException;
import com.fasterxml.jackson.databind.ObjectMapper;

/**
 * {@link DataTransformPolicyManager} class that overrides the methods in the
 * policy manager to add custom configurations for the data transform policies.
 * 
 * @author Booz Allen Hamilton
 *
 */
public class DataTransformPolicyManager extends AbstractPolicyManager {

    private static final Logger logger = LoggerFactory.getLogger(DataTransformPolicyManager.class);

    private static final ObjectMapper objectMapper = new ObjectMapper();

    private static DataTransformPolicyManager instance = null;

    /**
     * Returns the policy manager instance.
     * 
     * @return policy manager instance
     */
    public static DataTransformPolicyManager getInstance() {
        if (instance == null) {
            synchronized (DataTransformPolicyManager.class) {
                instance = new DataTransformPolicyManager();
            }
        }

        return instance;
    }

    /**
     * {@inheritDoc}
     */
    @Override
    public DefaultPolicy createPolicy(String policyIdentifier) {
        return new DefaultDataTransformPolicy(policyIdentifier);
    }

    /**
     * {@inheritDoc}
     */
    @Override
    public void setAdditionalConfigurations(Policy policy, PolicyInput input) {
        if (!DefaultDataTransformPolicy.class.isInstance(policy)) {
            throw new DataTransformException("Policy was not configured for data transform");
        }

        DefaultDataTransformPolicy dataTransformPolicy = (DefaultDataTransformPolicy) policy;

        // leveraging mash to handle mediator lookup based on policy rules
        BaseMediationManager mediationManager = dataTransformPolicy.getMediationManager();
        List<MediationConfiguration> mediationConfigurations = createMediationConfigurations(input.getRules());

        for (MediationConfiguration mediationConfiguration : mediationConfigurations) {
            mediationManager.validateAndAddMediator(mediationConfigurations, mediationConfiguration);
        }
    }

    private List<MediationConfiguration> createMediationConfigurations(List<PolicyRuleInput> ruleInputs) {
        List<MediationConfiguration> mediationConfigurations = new ArrayList<>();
        for (PolicyRuleInput ruleInput : ruleInputs) {
            MediationConfiguration mediationConfiguration = createMediationConfiguration(ruleInput);
            mediationConfigurations.add(mediationConfiguration);
        }

        return mediationConfigurations;
    }

    private MediationConfiguration createMediationConfiguration(PolicyRuleInput ruleInput) {
        MediationConfiguration mediationConfiguration = null;
        try {
            // convert rule configuration into mash's mediation configuration
            mediationConfiguration = objectMapper.convertValue(ruleInput.getConfigurations(),
                    MediationConfiguration.class);
            mediationConfiguration.setClassName(ruleInput.getClassName());

            logger.debug("Created mediation configuration: {}", mediationConfiguration);
        } catch (IllegalArgumentException e) {
            throw new DataTransformException("Invalid configurations found in rule", e);
        }

        return mediationConfiguration;
    }

}
